File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderSentenceSummaryTest.java
package uk.gov.justice.hmpps.prison.api.resource.impl;

import org.junit.jupiter.api.Test;
import org.springframework.core.ParameterizedTypeReference;
import uk.gov.justice.hmpps.prison.executablespecification.steps.AuthTokenHelper;

import static org.springframework.http.HttpMethod.GET;

public class OffenderSentenceSummaryTest extends ResourceTest {

    @Test
    public void getSentenceSummary() {

        final var token = authTokenHelper.getToken(AuthTokenHelper.AuthToken.VIEW_PRISONER_DATA);

        final var httpEntity = createHttpEntity(token, null);

        final var response = testRestTemplate.exchange(
            "/api/offenders/{nomsId}/booking/latest/sentence-summary",
            GET,
            httpEntity,
            new ParameterizedTypeReference<String>() {
            },
            "Z0020ZZ");

        assertThatJsonFileAndStatus(response, 200, "sentence_summary.json");
    }

}

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/OffenderSentenceSummaryTest.kt
package uk.gov.justice.hmpps.prison.api.resource.impl

import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@DisplayName("GET /api/offenders/{nomsId}/booking/latest/sentence-summary")
class OffenderSentenceSummaryTest : ResourceTest() {
  @Test
  fun `returns 401 without an auth token`() {
    webTestClient.get().uri("/api/offenders/Z0020ZZ/booking/latest/sentence-summary")
      .exchange()
      .expectStatus().isUnauthorized
  }

  @Test
  fun `should return 403 if does not have override role`() {
    webTestClient.get().uri("/api/offenders/Z0020ZZ/booking/latest/sentence-summary")
      .headers(setClientAuthorisation(listOf()))
      .exchange()
      .expectStatus().isForbidden
  }

  @Test
  fun `returns 403 when client has override role ROLE_SYSTEM_USER`() {
    webTestClient.get().uri("/api/offenders/Z0020ZZ/booking/latest/sentence-summary")
      .headers(setClientAuthorisation(listOf("ROLE_SYSTEM_USER")))
      .exchange()
      .expectStatus().isForbidden
  }

  @Test
  fun `returns success when client has override role ROLE_VIEW_PRISONER_DATA`() {
    webTestClient.get().uri("/api/offenders/Z0020ZZ/booking/latest/sentence-summary")
      .headers(setClientAuthorisation(listOf("ROLE_VIEW_PRISONER_DATA")))
      .exchange()
      .expectStatus().isOk
  }

  @Test
  fun `returns success if in user caseload`() {
    webTestClient.get().uri("/api/offenders/A1234AA/booking/latest/sentence-summary")
      .headers(setAuthorisation(listOf()))
      .exchange()
      .expectStatus().isOk
  }

  @Test
  fun `returns sentence summary data for the offender`() {
    webTestClient.get().uri("/api/offenders/Z0020ZZ/booking/latest/sentence-summary")
      .headers(setAuthorisation(listOf("ROLE_VIEW_PRISONER_DATA")))
      .exchange()
      .expectStatus().isOk
      .expectBody()
      .json("sentence_summary.json".readFile())
  }

  internal fun String.readFile(): String = this@OffenderSentenceSummaryTest::class.java.getResource(this).readText()
}
